File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        string curr = "";
        vector<string> res;
        solve(n, n, curr, res);

        return res;
    }

private:
    void solve(int l, int r, string curr, vector<string> &res) {
        if (l == 0 && r == 0) {
            res.push_back(curr);
            return;
        }
        
        if (l > 0) {
            curr += '(';
            solve(l - 1, r, curr, res);
            curr.pop_back();
        }
        

        if (r > l) {
            curr += ')';
            solve(l, r - 1, curr, res);
            curr.pop_back();
        }
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();

        solve(n, n, new StringBuilder(), res);
        return res;
    }

    private void solve(int l, int r, StringBuilder currStr, List<String> res) {
        // System.out.println("idx: " + idx + ", currStr: " + currStr);
        if (l == 0 && r == 0) {
            res.add(currStr.toString());
            return;
        }

        if (l > 0) {
            currStr.append('(');
            solve(l - 1, r, currStr, res);
            currStr.deleteCharAt(currStr.length() - 1);
        }
        
        if (r > l) {
            currStr.append(')');
            solve(l, r - 1, currStr, res);
            currStr.deleteCharAt(currStr.length() - 1);
        }
    }
}